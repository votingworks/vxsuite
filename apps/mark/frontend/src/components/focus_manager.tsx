import React, { useEffect, useRef } from 'react';
import { useLocation } from 'react-router-dom';
import styled from 'styled-components';
import { ScreenReader } from '../utils/ScreenReader';

const StyledFocusManager = styled.div`
  height: 100%;
  &:focus {
    outline: none;
  }
`;

export interface Props {
  children: React.ReactNode;
  onClick?: React.DOMAttributes<HTMLElement>['onClick'];
  onClickCapture?: React.DOMAttributes<HTMLElement>['onClickCapture'];
  onFocus?: React.DOMAttributes<HTMLElement>['onFocus'];
  onFocusCapture?: React.DOMAttributes<HTMLElement>['onFocusCapture'];
  onKeyDown?: React.DOMAttributes<HTMLElement>['onKeyDown'];
  onKeyDownCapture?: React.DOMAttributes<HTMLElement>['onKeyDownCapture'];
  screenReader: ScreenReader;
}

export function FocusManager({
  onKeyDown,
  onClick,
  onFocus,
  onKeyDownCapture,
  onClickCapture,
  onFocusCapture,
  children,
  screenReader,
}: Props): JSX.Element {
  const location = useLocation();
  const screen = useRef<HTMLDivElement>(null);
  useEffect(() => {
    function onPageLoad() {
      void screenReader.onPageLoad();

      // can't seem to find a better way than this, unfortunately.
      // the delay of 150 is to handle the case the Next button is selected
      // via arrow keys and then clicked. A shorter delay will fail to move
      // the focus away from "Next" in terms of audio. Even now, the Next button
      // stays highlighted, which is a bummer. We need to figure out a better solution.
      window.setTimeout(() => {
        const elementToFocus =
          document.getElementById('audiofocus') ?? screen.current;
        /* istanbul ignore next */
        if (!elementToFocus) return;
        elementToFocus.focus();
        elementToFocus.click();
      }, 150);
    }
    onPageLoad();
  }, [location.pathname, screenReader]);

  return (
    <StyledFocusManager
      ref={screen}
      tabIndex={-1}
      onKeyDown={onKeyDown}
      onClick={onClick}
      onFocus={onFocus}
      onKeyDownCapture={onKeyDownCapture}
      onClickCapture={onClickCapture}
      onFocusCapture={onFocusCapture}
    >
      {children}
    </StyledFocusManager>
  );
}
