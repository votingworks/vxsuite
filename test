#!/usr/bin/env bun

import { readFile } from 'fs/promises';
import { join } from 'path';

const { $ } = Bun;
export {};

const workspacePackageJson = await readFile(
  join(__dirname, 'package.json'),
  'utf8'
);
const { workspaces } = JSON.parse(workspacePackageJson);
const WIDTH = 100;

const results: { pkgPath: string; status: 'success' | 'failure' }[] = [];
let totalPassed = 0;
let totalFailed = 0;

for (const pkgPath of workspaces) {
  const tests = box(
    (async function* () {
      yield `→ bun run test --cwd ${pkgPath}`;
      const bunTest = $`bun run test --cwd ${pkgPath}`.quiet();

      const bunTestResult = await bunTest;

      yield* bunTestResult.stderr.toString('utf-8').split('\n');
      yield `${bunTestResult.exitCode === 0 ? '✅' : '❌'} ${pkgPath}`;

      if (bunTestResult.exitCode === 0) {
        results.push({ pkgPath, status: 'success' });
      } else {
        results.push({ pkgPath, status: 'failure' });
      }
    })()
  );

  let passed = 0;
  let failed = 0;

  for await (const line of tests) {
    const passedMatch = line.match(/(\d+) pass/);
    const failedMatch = line.match(/(\d+) fail/);

    if (passedMatch) {
      passed = parseInt(passedMatch[1], 10);
    }

    if (failedMatch) {
      failed = parseInt(failedMatch[1], 10);
    }

    console.log(line);
  }

  totalPassed += passed;
  totalFailed += failed;
}

console.log(
  `\n${totalPassed + totalFailed} tests: ${totalPassed} passed, ${totalFailed} failed`
);

for (const { pkgPath, status } of results) {
  console.log(`${status === 'success' ? '✅' : '❌'} ${pkgPath}`);
}

async function* box(lines: AsyncIterable<string>): AsyncIterable<string> {
  yield* lines;
}
