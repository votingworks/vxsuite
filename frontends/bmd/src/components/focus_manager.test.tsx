import React from 'react';
import { render } from '../../test/test_utils';

import { FocusManager } from './focus_manager';
import {
  AriaScreenReader,
  SpeechSynthesisTextToSpeech,
} from '../utils/ScreenReader';

it('renders FocusManager', () => {
  const { container } = render(
    <FocusManager
      screenReader={new AriaScreenReader(new SpeechSynthesisTextToSpeech())}
    >
      foo
    </FocusManager>
  );
  expect(container.firstChild).toMatchSnapshot();
});
