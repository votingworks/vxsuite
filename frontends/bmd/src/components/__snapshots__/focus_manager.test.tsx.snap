// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders FocusManager 1`] = `
.c0 {
  height: 100%;
}

.c0:focus {
  outline: none;
}

<div
  class="c0"
  tabindex="-1"
>
  foo
</div>
`;
